*** lib/pg_btree.c	2020-03-31 12:49:08.301500643 +0900
--- lib/pg_btree.c.PATCHED	2020-03-31 12:49:02.815773844 +0900
***************
*** 99,105 ****
  static void BTReaderReadPage(BTReader *reader, BlockNumber blkno);
  static IndexTuple BTReaderGetNextItem(BTReader *reader);
  
! static void _bt_mergebuild(Spooler *self, BTSpool *btspool);
  static void _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool,
  						  BTReader *btspool2, Relation heapRel);
  static int compare_indextuple(const IndexTuple itup1, const IndexTuple itup2,
--- 99,105 ----
  static void BTReaderReadPage(BTReader *reader, BlockNumber blkno);
  static IndexTuple BTReaderGetNextItem(BTReader *reader);
  
! static bool _bt_mergebuild(Spooler *self, BTSpool *btspool);
  static void _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool,
  						  BTReader *btspool2, Relation heapRel);
  static int compare_indextuple(const IndexTuple itup1, const IndexTuple itup2,
***************
*** 250,258 ****
  
  	for (i = 0; i < self->relinfo->ri_NumIndices; i++)
  	{
! 		if (spools[i] != NULL)
  		{
- 			_bt_mergebuild(self, spools[i]);
  			_bt_spooldestroy(spools[i]);
  		}
  		else
--- 250,257 ----
  
  	for (i = 0; i < self->relinfo->ri_NumIndices; i++)
  	{
! 		if (spools[i] != NULL && _bt_mergebuild(self, spools[i]))
  		{
  			_bt_spooldestroy(spools[i]);
  		}
  		else
***************
*** 379,386 ****
  	}
  }
  
! 
! static void
  _bt_mergebuild(Spooler *self, BTSpool *btspool)
  {
  	Relation heapRel = self->relinfo->ri_RelationDesc;
--- 378,384 ----
  	}
  }
  
! static bool
  _bt_mergebuild(Spooler *self, BTSpool *btspool)
  {
  	Relation heapRel = self->relinfo->ri_RelationDesc;
***************
*** 392,397 ****
--- 390,404 ----
  
  	tuplesort_performsort(btspool->sortstate);
  
+ 	{
+ 		const char* sort_method;
+ 		const char* space_type;
+ 		long	space_used;
+ 		tuplesort_get_stats(btspool->sortstate,
+ 			&sort_method, &space_type, &space_used);
+ 		if (space_type[0] == 'D')
+ 			return false;
+ 	}
  
  #if PG_VERSION_NUM >= 90300
  	/*
***************
*** 461,466 ****
--- 468,475 ----
  	}
  
  	BTReaderTerm(&reader);
+ 
+ 	return true;
  }
  
  /*
